### PR TITLE
synergy: fix x86_64-darwin build

### PR DESCRIPTION
**Pull request by mroi**
The Darwin/x86 build of `synergy` is broken since Qt5 was moved to the 11.0 Darwin SDK (commit [`5802c30`](https://github.com/NixOS/nixpkgs/tree/5802c306876a217179b0448613fbde4af8af1fae)). To unbreak `synergy`, it also needs to move to the 11.0 SDK itself.

## Description of changes

I changed the Darwin frameworks injected via `all-packages.nix`. In addition, I had to change the way the synergy package internally deals with the optional presence of the `UserNotifications` framework. The package should not pull the framework from `darwin` itself, but rely on the caller to either pass the right framework or not.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/misc/synergy/default.nix [hunk 1]
```diff
@@ -27,12 +27,12 @@
 , avahi-compat
 
   # MacOS / darwin
-, darwin
 , ApplicationServices
 , Carbon
 , Cocoa
 , CoreServices
 , ScreenSaver
+, UserNotifications ? null
```
Comment (wegank) [inline pkgs/applications/misc/synergy/default.nix:35]:
```suggestion
, UserNotifications
```

File: pkgs/applications/misc/synergy/default.nix [hunk 2]
```diff
@@ -50,7 +50,7 @@ stdenv.mkDerivation rec {
   patches = [
     # Without this OpenSSL from nixpkgs is not detected
     ./darwin-non-static-openssl.patch
-  ] ++ lib.optionals (stdenv.isDarwin && !(darwin.apple_sdk.frameworks ? UserNotifications)) [
+  ] ++ lib.optionals (stdenv.isDarwin && isNull UserNotifications) [
```
Comment (wegank) [inline pkgs/applications/misc/synergy/default.nix:53]:
```suggestion
  ] ++ lib.optionals stdenv.isDarwin [
```

File: pkgs/applications/misc/synergy/default.nix [hunk 3]
```diff
@@ -79,8 +79,8 @@ stdenv.mkDerivation rec {
     Cocoa
     CoreServices
     ScreenSaver
-  ] ++ lib.optionals (stdenv.isDarwin && darwin.apple_sdk.frameworks ? UserNotifications) [
-    darwin.apple_sdk.frameworks.UserNotifications
+  ] ++ lib.optionals (stdenv.isDarwin && !isNull UserNotifications) [
```
Comment (wegank) [inline pkgs/applications/misc/synergy/default.nix:82]:
```suggestion
  ] ++ lib.optionals stdenv.isDarwin [
```

File: pkgs/applications/misc/synergy/default.nix (hunk 2 above)
Comment (wegank) [reply pkgs/applications/misc/synergy/default.nix:53]: It's even better if the line can be removed.

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -35572,7 +35572,8 @@ with pkgs;
   };
 
   synergy = libsForQt5.callPackage ../applications/misc/synergy {
-    inherit (darwin.apple_sdk.frameworks) ApplicationServices Carbon Cocoa CoreServices ScreenSaver;
+    stdenv = if stdenv.isDarwin then darwin.apple_sdk_11_0.stdenv else stdenv;
+    inherit (darwin.apple_sdk_11_0.frameworks) ApplicationServices Carbon Cocoa CoreServices ScreenSaver;
```
Comment (wegank) [inline pkgs/top-level/all-packages.nix:35576]:
```suggestion
    inherit (darwin.apple_sdk_11_0.frameworks) ApplicationServices Carbon Cocoa CoreServices ScreenSaver UserNotifications;
```

File: pkgs/applications/misc/synergy/default.nix (hunk 1 above)
Comment (mroi) [reply pkgs/applications/misc/synergy/default.nix]: Done in commit 7bf7cb5a576fc72ccca5bb217fbb625e0c1a27af.

File: pkgs/applications/misc/synergy/default.nix (hunk 2 above)
Comment (mroi) [reply pkgs/applications/misc/synergy/default.nix]: Done in commit 7bf7cb5a576fc72ccca5bb217fbb625e0c1a27af.

File: pkgs/applications/misc/synergy/default.nix (hunk 3 above)
Comment (mroi) [reply pkgs/applications/misc/synergy/default.nix]: Done in commit 7bf7cb5a576fc72ccca5bb217fbb625e0c1a27af.

File: pkgs/top-level/all-packages.nix (hunk 4 above)
Comment (mroi) [reply pkgs/top-level/all-packages.nix:35576]: Done in commit 7bf7cb5a576fc72ccca5bb217fbb625e0c1a27af.

**Comment by mroi** [comment]
Thanks @wegank for the code review. I understand that you would prefer to simplify the synergy package by not allowing `UserNotifications` framework to be optional. Is my understanding correct?

I have implemented your suggested changes on my branch and will test them on Monday (only then do I have access to my synergy setup). I assume it is then preferred to roll all changes into one commit before merging?

**Comment by wegank** [comment]
> I understand that you would prefer to simplify the synergy package by not allowing UserNotifications framework to be optional

As the package now consistently uses `darwin.apple_sdk_11_0` instead of `darwin.apple_sdk` (which is 10.12 on x86_64-darwin, 11.0 on aarch64-darwin), `UserNotifications` always exists.

> I assume it is then preferred to roll all changes into one commit before merging?

It's preferred but not required, since I can click on the "squash and merge" button.

**Comment by mroi** [comment]
> As the package now consistently uses `darwin.apple_sdk_11_0` instead of `darwin.apple_sdk` (which is 10.12 on x86_64-darwin, 11.0 on aarch64-darwin), `UserNotifications` always exists.

Good point. Thanks for the explanation.

I will test the current patched version on Monday and then report here if everything works (which I would expect).

**Comment by mroi** [comment]
Latest commit tested. Works.

Compared to the last code review, the following has changed:
* disabled two tests on Darwin that would not succeed in the sandbox
* have the Darwin build somewhat respect the `withGUI` option (previously ignored on Darwin)
* rolled all changes into one commit

**Comment by wegank** [comment]
Result of `nixpkgs-review pr 283514` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>synergy</li>
    <li>synergyWithoutGUI</li>
  </ul>
</details>

**Pull request merged**